### PR TITLE
[ADD] : Display warning on cart if product out of stock

### PR DESCRIPTION
**Pull request by ajayjaviya**
Task : https://www.odoo.com/web#id=11457&view_type=form&model=project.task&menu_id=3942&action=327
Pad : https://pad.odoo.com/p/r.hjHDaFRnr7QNH3LH

### CODE REVIEW

File: addons/website_sale_stock/models/website.py [hunk 1]
```diff
@@ -0,0 +1,11 @@
+# -*- coding: utf-8 -*-
+from openerp import api, fields, models
+
+class Website(models.Model):
+    _inherit = 'website'
+
+    stock_warning_active =  fields.Boolean(compute="_compute_options", string='Out of Stock Warning')
```
Comment (jem-odoo) [inline addons/website_sale_stock/models/website.py]: Don't do that ! Execute your `_website_product_id_change` even if the view is not activated.

File: addons/website_sale_stock/models/sale_order.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+from openerp.addons.web.http import request
+
+class SaleOrder(models.Model):
+    _inherit = "sale.order"
+
+    @api.multi
+    def _website_product_id_change(self, order_id, product_id, qty=0, line_id=None):
+        self.ensure_one()
+        values = {}
+        if request.website.stock_warning_active:
```
Comment (jem-odoo) [inline addons/website_sale_stock/models/sale_order.py]: remove the condition

File: addons/website_sale/controllers/main.py [hunk 3]
```diff
@@ -272,8 +272,8 @@ def product(self, product, category='', search='', **kwargs):
 
         if not context.get('pricelist'):
             context['pricelist'] = int(self.get_pricelist())
-            product = template_obj.browse(cr, uid, int(product), context=context)
-
+            # Public user should able to read virtual stock
+            product = template_obj.browse(cr, SUPERUSER_ID, int(product), context=context)
```
Comment (jem-odoo) [inline addons/website_sale/controllers/main.py]: Is that really necessary ? Why should you read as SUPERUSER_ID for virtual stock ?

File: addons/website_sale/static/src/js/website_sale.js [hunk 4]
```diff
@@ -193,6 +193,7 @@ $('.oe_website_sale').each(function () {
             $product_id.val(0);
             $parent.find(".js_check_product").attr("disabled", "disabled");
         }
+        $product_id.trigger('change');
```
Comment (jem-odoo) [inline addons/website_sale/static/src/js/website_sale.js]: I guess this will trigger the js code of `website_sale_stock` ? 
Maybe you should make a odoo widget instead of native js code, and extend it in website_sale_stock.
Do it in a separate commit.

File: addons/website_sale_stock/views/website_sale_stock.xml [hunk 5]
```diff
@@ -1,5 +1,13 @@
 <?xml version="1.0" encoding="utf-8"?>
 <odoo>
+  <template id="assets_frontend" inherit_id="website.assets_frontend" name="Shop">
+      <xpath expr="." position="inside">
+        <script type="text/javascript" src="/website_sale_stock/static/src/js/website_sale_stock.js"></script>
+      </xpath>
+  </template>
+  <template id="products_out_of_stock_warning" inherit_id="website_sale.product" active="False" customize_show="True" name="Out of Stock Warning">
```
Comment (jem-odoo) [inline addons/website_sale_stock/views/website_sale_stock.xml]: empty template ?

File: addons/website_sale_stock/models/sale_order.py (hunk 2 above)
Comment (aja-odoo) [inline addons/website_sale_stock/models/sale_order.py]: then it will display warning message if customize option is not activate.

File: addons/website_sale_stock/views/website_sale_stock.xml (hunk 5 above)
Comment (aja-odoo) [inline addons/website_sale_stock/views/website_sale_stock.xml]: yes, to set customize option whether we need  to display (check stock) "out of stock warning" or not

File: addons/website_sale/controllers/main.py (hunk 3 above)
Comment (aja-odoo) [inline addons/website_sale/controllers/main.py]: yes, for public user we have to check available product qty

File: addons/website_sale_stock/models/website.py (hunk 1 above)
Comment (aja-odoo) [inline addons/website_sale_stock/models/website.py]: then it will display warning message if customize option is not activate.

File: addons/website_sale/static/src/js/website_sale.js (hunk 4 above)
Comment (jem-odoo) [inline addons/website_sale/static/src/js/website_sale.js]: Actually, setting the value of the input with javascript will not trigger automatically the change ? Do you really need to trigger it manually ?

Comment (aja-odoo) [inline addons/website_sale/static/src/js/website_sale.js]: yes , we need to trigger it , because when end user change the quantity we have to check that product is in stock or not based on that we display him/her a warning message.

File: addons/website_sale/controllers/main.py (hunk 3 above)
Comment (jem-odoo) [inline addons/website_sale/controllers/main.py]: I think to make the function which compute the `qty_available`of product as sudo, but I'm afraid it break lots of things, so for now, add access right for public for `stock.move, stock.warehouse, stock.quants, stock.location, product.uom`, and remove the SUPERUSER_ID above.

File: addons/website_sale_stock/views/website_sale_stock.xml [hunk 6]
```diff
@@ -31,4 +39,35 @@
...
+  </template>
+  <template id="product_quantity_out_of_stock" inherit_id="website_sale.product_quantity" name="Set Stock Limit">
+    <xpath expr="//input[@name='add_qty']" position="replace">
+        <input type="text" t-attf-class="form-control #{'js_check_stock' if website.stock_warning_active and product.type=='product' else ''}" data-min="1" name="add_qty" value="1"/>
+    </xpath>
+  </template>
+
+  <template id="product_variants_out_of_stock" inherit_id="website_sale.product_variants" name="Out of Stock Product Variants">
+    <xpath expr="//input[@class='js_product_change']" position="replace">
+        <input type="radio" name="product_id" t-attf-class="js_product_change #{ 'js_check_stock' if website.stock_warning_active else '' }" t-att-checked="'checked' if variant_id_index == 0 else ''" t-att-value="variant_id.id" t-att-data-product="variant_id.id" t-att-data-lst_price="variant_id.lst_price" t-att-data-price="variant_id.price"/>
```
Comment (jem-odoo) [inline addons/website_sale_stock/views/website_sale_stock.xml]: don't add `js_check_stock`, you can find it with another selector, using the class already on the input tag.

File: addons/website_sale_stock/views/website_sale_stock.xml [hunk 7]
```diff
@@ -31,4 +39,35 @@
...
+  <template id="products_out_of_stock" inherit_id="website_sale.products_add_to_cart" name="Out of Stock Products">
+    <xpath expr="//a[@class='btn btn-default btn-xs a-submit']" position="replace">
+        <a t-attf-class="btn btn-default btn-xs a-submit #{ 'disabled' if website.stock_warning_active and product.qty_available&lt;= 0 and product.type=='product' else '' }">
+            <span class="fa fa-shopping-cart"/>
+        </a>
+    </xpath>
+  </template>
+
+  <template id="product_out_of_stock" inherit_id="website_sale.product" name="Out of Stock Product">
+    <xpath expr="//a[@id='add_to_cart']" position="replace">
```
Comment (jem-odoo) [inline addons/website_sale_stock/views/website_sale_stock.xml]: don't replace the complete button. Just add your input with `data-stock` after the input `name="product_id"`.

File: addons/website_sale_stock/views/website_sale_stock.xml (hunk 5 above)
Comment (jem-odoo) [inline addons/website_sale_stock/views/website_sale_stock.xml]: This template should be the only `customize_show` : it should add the `data-stock`input tag, and the warning message (`t-if='product.qty_available > 0 ' .... `).

File: addons/website_sale_stock/views/website_sale_stock.xml [hunk 8]
```diff
@@ -31,4 +39,35 @@
       </t>
     </xpath>
   </template>
-</odoo>
\ No newline at end of file
+
+  <template id="products_out_of_stock" inherit_id="website_sale.products_add_to_cart" name="Out of Stock Products">
```
Comment (jem-odoo) [inline addons/website_sale_stock/views/website_sale_stock.xml]: remove it

**Comment by jem-odoo** [comment]
Some ideas : 
For /shop/product, you should have one template (customize_show) to add the warning message, and the data required for Javascript, used when switching variants and changing the quantity.

For shop/cart, the warning message should be done in javascript and dynamically (normally, when you arrive on the cart page, you should have the message, since the 'add to cart' button is disabled when you reach the limit). Extends the 'cart' template to add a `data-stock` attribute on the input (with js_quantity class), and check if the `input content >= data-stock`, then display warning message. No code in the controller.

File: addons/website_sale_stock/views/website_sale_stock.xml (hunk 6 above)
Comment (aja-odoo) [inline addons/website_sale_stock/views/website_sale_stock.xml]: i add `js_check_stock` only  if customize option is true `(website.stock_warning_active)` , if we use other existing selector then we don't know customize options is enable or disable

File: addons/website_sale/controllers/main.py (hunk 3 above)
Comment (aja-odoo) [inline addons/website_sale/controllers/main.py]: done

File: addons/website_sale_stock/views/website_sale_stock.xml (hunk 5 above)
Comment (aja-odoo) [inline addons/website_sale_stock/views/website_sale_stock.xml]: if we do it , then we cant render message on variant change eg. if product with color "white" have stock `0`  quantity, then we need to display message `Sorry ! This product is out of stock` , and now user change same product's variant color "black" have stock `5` quantity and user ask for `10` quantity then we have to display `Sorry ! Only 5 units are still in stock` and if user change variant "red" of same product have `100` quantity and end user ask for `10` then no message and he can proceed further.
